book/src/rfcs/benchmarking.md

```diff
@@ -0,0 +1,266 @@
+hardest to measure, because the environment can have a considerable effect over
+the results. Some high quality machine would be required, ideally a fully
+dedicated one to minimize noise. Further in this document, we mention a way of
+measuring that overcomes noise problem.  It can be subdivided into 2 other
+metrics or alternative approaches:
+
+: Execution time The simplest approach. It runs the program several times and
+returns the overall average execution time. This benchmark might become slow in
+cases where the program takes too much time to run. In such scenarios, this
+benchmark can be made faster by reducing the number of samples (i.e., worst
+precision) it uses. This is one of the most interesting metrics for external
+users and a key datapoint to compare how our proof-system fares against
+competitors. The execution time metric can also help our engineering teams as a
+metric to identify regressions or to verify performance improvements.
+
+: CPU instructions Instead of recording execution time, this method counts the
+number of cpu instructions used in a program execution, not very useful for
+external developer given that it doesn't translate directly to time. It is
+mostly useful to our engineering teams as an alternative to execution time, it
+is faster given that you only have to run it once to collect the information,
+and can be run on cheap machines without significative noise given that
+execution time doesn't affect the number of instructions. Additionally, the
+tools used to measure it also collect some other metrics that can be useful to
+aide optimization like cache misses and branch mispredictions.
+
+#### Memory
+
+Memory usage is another interesting metric. There are different ways of
+measuring memory but the most useful for us is peak memory, as this will tell
+whether something can even be used in a particular environment. It can also help
+finding errors related to unusual memory usage.  Memory usage is easier to
+measure, as it does not appear to depend as much on the noise of the environment
+as long as there is enough memory.
+
+#### Hardware utilization
+
+It can be of interest in some cases to measure how much of the available
+resources are actually used, ideally the same computer will take half the time
+with twice the resources, unfortunately this isn't always the case and sometimes
+additional resources go unused.  This doesn't change that often, the main
+utility of these benchmarks is just to detect low utilization so that it can be
+improved.  Hardware utilization could include many things, GPU acceleration for
+example if we were to start using it, but for now will only include CPU/core
+utilization.
+
+### Layers
+
+There are 3 main layers of the stack that we want to consider, kimchi, snarky
+and pickles. Given that they are implemented in different languages and exist in
+different repositories special considerations each layer will be required.
+
+## Where to run the benchmarks
```

Comment:
Can we run the benchmarks within the profiling frontend that Viable Systems created? https://perf.ci.openmina.com/builds . I bet the velocity team can help support this. I’d love to see explanations on how that would work in this RFC and I’d recommend looking into before deciding on benchmarking libraries because I know they’ve already instrumented some of the rust and ocaml code. I’ll invite someone from viable systems to comment here to describe how it works from their side.